File: Leetcode 22 - generate-parentheses/22.generate-parentheses.cpp
/*
 * @lc app=leetcode id=22 lang=cpp
 *
 * [22] Generate Parentheses
 */

// @lc code=start
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
  void generateParenthesisHelper(int open, int close, string s,
                                 vector<string> &result) {
    if (open == 0 && close == 0) {
      result.push_back(s);
      return;
    }

    if (open > 0) {
      string cur = s + '(';
      generateParenthesisHelper(open - 1, close, cur, result);
    }

    if (open < close) {
      string cur = s + ')';
      generateParenthesisHelper(open, close - 1, cur, result);
    }
  }
  vector<string> generateParenthesis(int n) {
    vector<string> result;
    string s = "";
    generateParenthesisHelper(n, n, s, result);
    return result;
  }
};

int main() {
  Solution sol;

  vector<string> result = sol.generateParenthesis(3);
  for (string s : result) {
    cout << s << endl; // ((())) (()()) (())() ()(()) ()()() (()) (()()) ((()))
  }

  result = sol.generateParenthesis(1);
  for (string s : result) {
    cout << s << endl; // ()
  }
}
// @lc code=end

File: Leetcode 22 - generate-parentheses/22.generate-parentheses.java
/*
 * @lc app=leetcode id=22 lang=java
 *
 * [22] Generate Parentheses
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {

  public List<String> generateParenthesis(int n) {
    List<String> result = new ArrayList<>((int) Math.pow(2, n));
    helper(n, n, result, new StringBuilder(2 * n));
    return result;
  }

  private void helper(
    int open,
    int close,
    List<String> result,
    StringBuilder sb
  ) {
    if (open == 0 && close == 0) {
      result.add(sb.toString());
      return;
    }

    if (open > 0) {
      sb.append('(');
      helper(open - 1, close, result, sb);
      sb.setLength(sb.length() - 1);
    }

    if (open < close) {
      sb.append(')');
      helper(open, close - 1, result, sb);
      sb.setLength(sb.length() - 1);
    }
  }

  public static void main(String[] args) {
    Solution solution = new Solution();

    System.out.println(solution.generateParenthesis(3)); // ["((()))","(()())","(())()","()(())","()()()"]

    System.out.println(solution.generateParenthesis(1)); // ["()"]

    System.out.println(solution.generateParenthesis(2)); // ["(())","()()"]
  }
}
// @lc code=end
